src/types.rs

```diff
@@ -350,6 +350,29 @@ impl<N: Name> Type<N> {
             _ => None,
         }
     }
+    /// If the type is an arrow, recursively get all curried function arguments.
+    pub fn args_destruct(self) -> Option<Vec<Type<N>>> {
+        match self {
+            Type::Constructed(n, mut args) if n.is_arrow() => {
+                let mut tps = Vec::with_capacity(1);
+                args.reverse();
+                tps.push(args.pop().unwrap());
+                let mut tp = args.pop().unwrap();
+                loop {
+                    match tp {
+                        Type::Constructed(n, mut args) if n.is_arrow() => {
+                            args.reverse();
```

Comment:
small perf thing but here and above you can avoid the `reverse()` calls by swapping the order of of `tps.push` and `tp=`

Reply:
Nice!